opta/cloud/aws/state.py

```diff
@@ -0,0 +1,227 @@
+        return True
+
+    def read_raw(self) -> str:
+        bucket = self._bucket
+        region = self._region
+        key = self._object_key
+        logger.debug(
+            f"Found an s3 backend in bucket {bucket} and key {key}, "
+            "gonna try to download the statefile from there"
+        )
+
+        s3 = _client.s3(region)
+
+        f = io.StringIO()
+        try:
+            s3.download_fileobj(Bucket=bucket, Key=key, Fileobj=f)
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "404":
+                logger.debug("Did not find terraform state file")
+                raise StateNotFoundError()
+
+            raise
+
+        contents = f.getvalue()
+
+        return contents
+
+    @property
+    def _bucket(self) -> str:
+        c = self.config
+
+        return f"opta-tf-state-{c.org_name}-{c.layer_name}"
+
+    def _configure_bucket(self) -> None:
+        bucket_name = self._bucket
+        region = self._region
+
+        s3 = _client.s3(region)
+
+        try:
+            s3.get_bucket_encryption(Bucket=bucket_name)
+            return
+        except ClientError as e:
+            error_code = e.response["Error"]["Code"]
+            if error_code == "AuthFailure":
+                raise UserErrors(
+                    "The AWS Credentials are not configured properly.\n"
+                    "Visit https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/setup-credentials.html "
+                    "for more information."
+                )
+
+            if error_code == "AccessDenied":
+                raise UserErrors(
+                    f"We were unable to access the S3 bucket, {bucket_name} on your AWS account (opta needs this to store state).\n"
+                    "Possible Issues: \n"
+                    " - Bucket name is not unique and might be present in some other Account. Try updating the name in Configuration file to something else.\n"
+                    " - It could also mean that your AWS account has insufficient permissions.\n"
+                    "Please fix these issues and try again!"
+                )
+
+            if error_code != "NoSuchBucket":
+                error_message = e.response["Error"]["Message"]
+                raise UserErrors(
+                    "When trying to determine the status of the state bucket, we got an "
+                    f"{error_code} error with the message {error_message}"
+                )
+
+        logger.debug("S3 bucket for terraform state not found, creating a new one")
+
+        create_kwargs: Dict[str, Any] = {}
+        if region != "us-east-1":
+            location = cast(BucketLocationConstraintType, region)
+            create_kwargs["CreateBucketConfiguration"] = {"LocationConstraint": location}
+
+        s3
+        s3.create_bucket(Bucket=bucket_name, **create_kwargs)
+        time.sleep(
+            10
+        )  # TODO: Change this for a "wait until bucket resource is available"
```

Comment:
possibly this https://stackoverflow.com/a/49817544
ok with sleep for now if you think that's enough time